### DISCUSSION TITLE
Add datetime filtering field in the mosaic endpoint

### DISCUSSION THREAD

**Discussion by charalamm** [developmentseed/eoAPI#137] [Ideas]
Hello,

We have started moving closer and closer to production with a project that uses eoAPI and I have noticed a thing that would be useful for us.

We have one product for many dates and we want to show mosaics for each date on a frontend app. As far as I understand, for the mosaic endpoint to work we need to register a search. Since we have products for different dates, we must register a new search for each available date in order to show mosaics of different dates. In my opinion, it would be useful to be able to add a datetime field in the mosaic endpoint and get the mosaic for the specified date. What do you think?

**Comment by vincentsarago** [comment]
@charalamm For now we don't have really other choice but need to `register` mosaics before being able to create tiles (that's how titiler-pgstac works). 

But you could add custom code that register the mosaic on each endpoint request to get the `searchId` so basically have endpoints like `/collections/{collection_id}/datetime/{datetime}`

I could maybe write some pseudo code but that could take some time

**Reply by charalamm** [reply]
Thank you @vincentsarago for your quick response. If you are interested in this feature I am available to work on it with some guidance. However I see that a clean solution would require to change the pgstac extension and I am not sure if it is wise to got there

**Reply by vincentsarago** [reply]
@charalamm I don't think this is a feature that will be needed for most of the user of this demo application to be honest. I've proposed something similar in titiler-pgstac but only for collection (no property filter) https://github.com/stac-utils/titiler-pgstac/issues/118 but right now titler-pstac development is kinda on pause because we need $$ or external contributors.

**Reply by vincentsarago** [reply]
👋 @charalamm I've started a PR in titiler-pgstac to kinda enable this

https://github.com/stac-utils/titiler-pgstac/pull/127

Once/If this gets merged we should be (in theory) able to do something like:

```python
def CollectionIdDatetimeParams(
    collection_id: Annotated[
        str,
        Path(description="Collection Id"),
    ],
    datetime: Annotated[
        str,
        Path(description="datetime"),
    ]
    connection=Depends(db_conn),
) -> str:
    """SearchId"""
    search = model.PgSTACSearch(
        filter={
            "op": "and",
            "args": [
                {
                    "op": "=",
                    "args": [{"property": "collection"}, collection_id],
                },
                {
                    "op": "=",
                    "args": [{"property": "datetime"}, datetime],
                }
            ],
        },
        filter_lang="cql2-json"
    )

    metadata = model.Metadata(name=f"Mosaic for {collection_id} Collection at {datetime}")

    with connection.cursor(row_factory=class_row(model.Search)) as cursor:
        cursor.execute(
            "SELECT * FROM search_query(%s, _metadata => %s);",
            (
                search.model_dump_json(by_alias=True, exclude_none=True),
                metadata.model_dump_json(exclude_none=True),
            ),
        )
        search_info = cursor.fetchone()

    return search_info.id




###############################################################################
# Collection Endpoints
collections_datetime = MosaicTilerFactory(
    path_dependency=CollectionIdDatetimeParams,
    router_prefix="/collections/{collection_id}/datetime/{datetime}",
)
app.include_router(collections_datetime.router, tags=["Collection"], prefix="/collections/{collection_id}/datetime/{datetime}")
```

**Reply by charalamm** [reply]
Hi @vincentsarago. Thank you for looking into this! I am sorry for not replying all this time. I was really busy.
I see that the PR is closed, but that is ok. I appreciate you putting time into this!

Back then I ended up doing some changes in a [fork](https://github.com/stac-utils/titiler-pgstac/compare/main...DHI:titiler-pgstac:main) of titiler.pgstac and in a fork of [pgstac](https://github.com/stac-utils/pgstac/compare/main...DHI:pgstac:main). However they are outdated now.

 Anyway, since the register request is so fast we might go in that way to be in accordance with your repo.

**Comment by sharkinsspatial** [comment]
@charalamm The mosaic registration call is very light.  Internally, titiler-pgstac is using a hash of the filter body to generate the search id which is stored and returned to the client.  So subsequent mosaic registration calls with the same search body return the same result.  A frontend pattern that we have used on several NASA applications is that when a user selects a date, the frontend makes the mosaic registration call with a search for that date and the returned id is used to request tiles for that date from the tiler endpoint.  We have a few examples of applications with a time slider component that use this pattern.

**Reply by charalamm** [reply]
Thanks @sharkinsspatial. That is insightful. We might go for it. Are there any advantages to using this approach over adding a datetime filter in the request (without taking into consideration the development time which of course is very important)?

**Comment by vincentsarago** [comment]
FYI I just merged https://github.com/stac-utils/titiler-pgstac/pull/132 which add `/collections/{collection_id}` endpoints. 

The `magic` is being done by the [factory prefix](https://github.com/stac-utils/titiler-pgstac/blob/66cb0d5b52a7ac0b5fbf2b7c1d4b3f742c98ed99/titiler/pgstac/main.py#L159) and the custom `path_dependency` https://github.com/stac-utils/titiler-pgstac/blob/66cb0d5b52a7ac0b5fbf2b7c1d4b3f742c98ed99/titiler/pgstac/dependencies.py#L37-L91

In theory you can have could also add other filter as path or query-parameters to create Mosaic

**Reply by charalamm** [reply]
Ok, I will make an attempt to use it. Thanks!
